packages/medusa/src/strategies/batch-jobs/product/export.ts

```diff
@@ -0,0 +1,420 @@
+          .retrieve(batchJobId)) as ProductExportBatchJob
+
+        const { writeStream, fileKey, promise } = await this.fileService_
+          .withTransaction(transactionManager)
+          .getUploadStreamDescriptor({
+            name: "product-export",
+            ext: "csv",
+          })
+
+        const header = await this.buildHeader(batchJob)
+        writeStream.write(header)
+
+        advancementCount =
+          batchJob.result?.advancement_count ?? advancementCount
+        offset = (batchJob.context?.list_config?.skip ?? 0) + advancementCount
+        limit = batchJob.context?.list_config?.take ?? this.BATCH_SIZE
+
+        const { list_config = {}, filterable_fields = {} } = batchJob.context
+        const [productList, count] = await this.productService_
+          .withTransaction(transactionManager)
+          .listAndCount(filterable_fields, {
+            ...list_config,
+            skip: offset,
+            take: limit,
+          } as FindProductConfig)
+
+        productCount = count
+        let products: Product[] = productList
+
+        while (offset < productCount) {
+          if (!products?.length) {
+            products = await this.productService_
+              .withTransaction(transactionManager)
+              .list(filterable_fields, {
+                ...list_config,
+                skip: offset,
+                take: limit,
+              } as FindProductConfig)
+          }
+
+          products.forEach((product: Product) => {
+            const lines = this.buildProductVariantLines(product)
+            lines.forEach((line) => writeStream.write(line))
+          })
+
+          advancementCount += products.length
+          offset += products.length
+          products = []
+
+          batchJob = (await this.batchJobService_
+            .withTransaction(transactionManager)
+            .update(batchJobId, {
+              result: {
+                file_key: fileKey,
+                count: productCount,
+                advancement_count: advancementCount,
+                progress: advancementCount / productCount,
+              },
+            })) as ProductExportBatchJob
+
+          if (batchJob.status === BatchJobStatus.CANCELED) {
+            writeStream.end()
+
+            await this.fileService_
+              .withTransaction(transactionManager)
+              .delete({ key: fileKey })
+            return
+          }
+        }
+
+        writeStream.end()
+
+        return await promise
+      },
+      "REPEATABLE READ",
+      async (err) =>
+        this.handleProcessingError(batchJobId, err, {
+          count: productCount,
+          advancement_count: advancementCount,
+          progress: advancementCount / productCount,
+        })
+    )
+  }
+
+  public async buildHeader(batchJob: ProductExportBatchJob): Promise<string> {
+    const {
+      prices = [],
+      dynamicImageColumnCount,
+      dynamicOptionColumnCount,
+    } = batchJob?.context?.shape ?? {}
```

Comment:
Praise: this is very nice and readable, solid push to move the computation of these into the preprocess method